### PR TITLE
DSP: Implement audio filters (simple, biquad)

### PR DESCRIPTION
**Pull request by MerryMage**
In implementing this I went for ease of understanding. Modern compilers probably do a good job optimizing things like this anyway.

Matches firmware output exactly.

### CODE REVIEW

File: src/audio_core/hle/filter.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+void SourceFilters::Configure(SourceConfiguration::Configuration::BiquadFilter config) {
+    biquad_filter.Configure(config);
+}
+
+void SourceFilters::ProcessFrame(StereoFrame16& frame) {
+    if (!simple_filter_enabled && !biquad_filter_enabled)
+        return;
+
+    for (auto& x : frame) {
+        if (simple_filter_enabled)
```
Comment (lioncash) [inline src/audio_core/hle/filter.cpp]: you can probably make this

```cpp
if (simple_filter_enabled) {
    std::transform(frame.begin(), frame.end(), frame.begin(),
                   [this](const auto& data) { return simple_filter.ProcessSample(data); });
}

if (biquad_filter_enabled) {
    std::transform(frame.begin(), frame.end(), frame.begin(),
                   [this](const auto& data) { return biquad_filter.ProcessSample(data); });
}
```
(possibly extracting the inner code out into a helper somehow) which avoids doing redundant compares in a loop.

File: src/audio_core/hle/filter.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+// BiquadFilter
+
+void SourceFilters::BiquadFilter::Reset() {
+    x1.fill(0);
+    x2.fill(0);
+    y1.fill(0);
+    y2.fill(0);
+    // Configure as passthrough.
+    a1 = a2 = b1 = b2 = 0;
```
Comment (lioncash) [inline src/audio_core/hle/filter.cpp]: Is there any particular reason aX and bX aren't just arrays?

Comment (MerryMage) [inline src/audio_core/hle/filter.cpp]: b0 doesn't exist.

File: src/audio_core/hle/filter.cpp (hunk 1 above)
Comment (MerryMage) [inline src/audio_core/hle/filter.cpp]: I'll actually be doing this in a few more places so I'll pop that in common.h, good idea.

File: src/audio_core/hle/common.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+/// The DSP is quadraphonic internally.
+using QuadFrame32   = std::array<std::array<s32, 4>, AudioCore::samples_per_frame>;
+
+/**
+ * This performs the filter operation defined by FilterT::ProcessSample on the frame in-place.
+ * FilterT::ProcessSample is called sequentially on the samples.
+ */
+template<typename FrameT, typename FilterT>
+void FilterFrame(FrameT& frame, FilterT& filter) {
+    std::transform(frame.begin(), frame.end(), frame.begin(), [&filter](const typename FrameT::value_type& sample) {
```
Comment (lioncash) [inline src/audio_core/hle/common.h]: is there any particular reacson `const auto& sample` isn't just used? It's effectively the same thing.

**Comment by JayFoxRox** [comment]
I didn't try compiling/running but the code looks fine (Format and algorithmicly).

One question though: Is the `<< 14` / `<< 15` from RE'ing the firmware or did you just choose these for the appropiate precision? (and do those work for all input values if you chose them yourself?)

Other than that, code LGTM

**Comment by MerryMage** [comment]
Shifts are from observation of firmware behaviour, verified ([here are some simplified tests](https://github.com/MerryMage/3ds-audio/)) and produces results exactly equivalent for all values on the firmware revisions I've tested.

Background trivia: When writing DSP firmware it is typical to use fixed point math for anything that touches high frequency stuff because floating point (used to be) too slow. This is one of those things people [love optimizing](http://www.infineon.com/dgdl/sj2004284[1].pdf?fileId=db3a304412b407950112b40f7c7e13fd).

File: src/audio_core/hle/filter.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    /// Reset internal state.
+    void Reset();
+
+    /**
+     * Enable/Disable filters
+     * See also: SourceConfiguration::Configuration::simple_filter_enabled,
+     *           SourceConfiguration::Configuration::biquad_filter_enabled.
+     */
+    void Enable(bool simple, bool biquad);
```
Comment (bunnei) [inline src/audio_core/hle/filter.h]: Missing docstring descriptions for arguments (ditto to the below methods)

File: src/audio_core/hle/filter.h [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        s32 a1, a2, b0, b1, b2;
+        // Internal state
+        std::array<s16, 2> x1;
+        std::array<s16, 2> x2;
+        std::array<s16, 2> y1;
+        std::array<s16, 2> y2;
+    } biquad_filter;
+};
+
+}
```
Comment (bunnei) [inline src/audio_core/hle/filter.h]: Nitpick, but it's nice to add a `} // namespace blah` comment these for readability

File: src/audio_core/hle/filter.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
```
Comment (bunnei) [inline src/audio_core/hle/filter.h]: Ditto

File: src/audio_core/hle/filter.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+    x2 = x1;
+    x1 = x0;
+    y2 = y1;
+    y1 = y0;
+
+    return y0;
+}
+
+}
```
Comment (bunnei) [inline src/audio_core/hle/filter.cpp]: Nitpick, but it's nice to add a `} // namespace blah` comment for readability

File: src/audio_core/hle/filter.cpp [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+}
```
Comment (bunnei) [inline src/audio_core/hle/filter.cpp]: Ditto

**Comment by bunnei** [comment]
I don't really know how this stuff works, so I assume you tested it and it is right. From just a general code review, LGTM, other than the minor comment nits.

**Comment by MerryMage** [comment]
Updated as per @bunnei's comments above.

**Comment by JamePeng** [comment]
LGTM

File: src/audio_core/hle/common.h (hunk 3 above)
Comment (lioncash) [inline src/audio_core/hle/common.h]: Since we compile as C++14 now, the argument can just be `const auto& sample`

Comment (MerryMage) [inline src/audio_core/hle/common.h]: Noted. I'll include that in my next PR.

**Pull request merged**